src/audio/dai-zephyr.c

```diff
@@ -848,12 +848,38 @@ static int dai_set_dma_buffer(struct dai_data *dd, struct comp_dev *dev,
 
 	/* calculate DMA buffer size */
 	period_count = dd->dma->plat_data.period_count;
+
+#if CONFIG_IPC_MAJOR_4
+	/* Use ibs/obs from copier gtw config to calculate DMA buffer size in ms.
+	 * This is used to calculate correct period count value.
+	 * It is especially important for FW aggregation use case, when Copier
+	 * receives whole DMA buffer size in GTW config but should be able to
+	 * allocate multiple DMA's implicitly.
+	 * E.g. if dma_buffer_size received in Copier GTW config is 7680 bytes and FW
+	 * should aggregate two streams 1ch/24_32bit/48kHz and 3ch/24_32bit/48kHz into one
+	 * 4ch/24_32bit/48kHz stream, then FW will have to allocate two DMA's with sizes
+	 * 1920 and 5760 bytes respectively and same 10 ms period for both to avoid
+	 * glitches.
+	 */
+	struct ipc4_copier_module_cfg *copier_cfg = dd->dai_spec_config;
+	uint32_t dma_buff_ms_size;
+
+	if (dev->direction == SOF_IPC_STREAM_CAPTURE)
+		dma_buff_ms_size = dd->ipc_config.dma_buffer_size / copier_cfg->base.ibs;
+	else
+		dma_buff_ms_size = dd->ipc_config.dma_buffer_size / copier_cfg->base.obs;
+
+	period_count = MAX(period_count, dma_buff_ms_size);
```

Comment:
Not quite following the logic. Period size may not always be 1ms, so how come we mix "period_count" and "dma_buff_ms_size" here? Or is the assumption that copier "ibs" must be aligned to period size of the pipeline. This would be good to capture in a comment. The variable name still needs to be corrected, it needs to be "dma_buff_size_in_periods" or something like that.

Reply:
ibs/obs size depends on module processing domain, in case of Copier(LL) period will be 1ms. In Copier gtw config we receive DMA size aligned with ibs/obs. Based on previous assumptions we can calculate DMA buffer length in ms, which is also DMA `period_count`.  I agree the name of the variable and comment should be changed.